# Taller Análisis Exploratorio de Datos

## Paso 1: Carga de datos

Abra el archivo en Excel o bloc de notas, y revise sí hay algo particular con los datos. Este paso es importante para evitar problemas posteriores al cargar los datos.

Cargar el archivo dado, que contiene un dataset cuya información puede consultar en [IBM](https://community.ibm.com/community/user/businessanalytics/blogs/steven-macko/2019/07/11/telco-customer-churn-1113).

Responda las siguientes preguntas:

- ¿Cuántas filas y columnas tiene el dataframe?
- ¿De qué tipo de dato es cada columna?
- ¿Las variables están en el tipo de dato correcto para su contenido?


In [12]:
import pandas as pd
df = pd.read_csv("Telco_customer_churn.csv")
df

,CustomerID,Count,Country,State,City,Zip Code,Lat Long,Latitude,Longitude,Gender,...,Contract,Paperless Billing,Payment Method,Monthly Charges,Total Charges,Churn Label,Churn Value,Churn Score,CLTV,Churn Reason
0,3668-QPYBK,1,United States,California,Los Angeles,90003,"33.964131, -118.272783",33.964131,-118.272783,Male,...,Month-to-month,Yes,Mailed check,53.85,108.15,Yes,1,86,3239,Competitor made better offer
1,9237-HQITU,1,United States,California,Los Angeles,90005,"34.059281, -118.30742",34.059281,-118.307420,Female,...,Month-to-month,Yes,Electronic check,70.70,151.65,Yes,1,67,2701,Moved
2,9305-CDSKC,1,United States,California,Los Angeles,90006,"34.048013, -118.293953",34.048013,-118.293953,Female,...,Month-to-month,Yes,Electronic check,99.65,820.5,Yes,1,86,5372,Moved
3,7892-POOKP,1,United States,California,Los Angeles,90010,"34.062125, -118.315709",34.062125,-118.315709,Female,...,Month-to-month,Yes,Electronic check,104.80,3046.05,Yes,1,84,5003,Moved
4,0280-XJGEX,1,United States,California,Los Angeles,90015,"34.039224, -118.266293",34.039224,-118.266293,Male,...,Month-to-month,Yes,Bank transfer (automatic),103.70,5036.3,Yes,1,89,5340,Competitor had better devices
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
7038,2569-WGERO,1,United States,California,Landers,92285,"34.341737, -116.539416",34.341737,-116.539416,Female,...,Two year,Yes,Bank transfer (automatic),21.15,1419.4,No,0,45,5306,NaN
7039,6840-RESVB,1,United States,California,Adelanto,92301,"34.667815, -117.536183",34.667815,-117.536183,Male,...,One year,Yes,Mailed check,84.80,1990.5,No,0,59,2140,NaN
7040,2234-XADUH,1,United States,California,Amboy,92304,"34.559882, -115.637164",34.559882,-115.637164,Female,...,One year,Yes,Credit card (automatic),103.20,7362.9,No,0,71,5560,NaN
7041,4801-JZAZL,1,United States,California,Angelus Oaks,92305,"34.1678, -116.86433",34.167800,-116.864330,Female,...,Month-to-month,Yes,Electronic check,29.60,346.45,No,0,59,2793,NaN


¿Cuántas filas y columnas tiene el dataframe?

In [13]:
f"filas: {df.shape[0]}"

'filas: 7043'

In [14]:
f"columnas: {df.shape[1]}"

'columnas: 33'

¿De qué tipo de dato es cada columna?

In [15]:
df.dtypes

CustomerID               str
Count                  int64
Country                  str
State                    str
City                     str
Zip Code               int64
Lat Long                 str
Latitude             float64
Longitude            float64
Gender                   str
Senior Citizen           str
Partner                  str
Dependents               str
Tenure Months          int64
Phone Service            str
Multiple Lines           str
Internet Service         str
Online Security          str
Online Backup            str
Device Protection        str
Tech Support             str
Streaming TV             str
Streaming Movies         str
Contract                 str
Paperless Billing        str
Payment Method           str
Monthly Charges      float64
Total Charges            str
Churn Label              str
Churn Value            int64
Churn Score            int64
CLTV                   int64
Churn Reason             str
dtype: object

¿Las variables están en el tipo de dato correcto para su contenido?

Total Charges debería de ser flotante

In [16]:
df["Total Charges"] =  pd.to_numeric(df["Total Charges"], errors='coerce')

Lo mismo pasa con "Churn Value"

In [17]:
df["Churn Value"] =  pd.to_numeric(df["Churn Value"], errors='coerce')

In [18]:
df.dtypes

CustomerID               str
Count                  int64
Country                  str
State                    str
City                     str
Zip Code               int64
Lat Long                 str
Latitude             float64
Longitude            float64
Gender                   str
Senior Citizen           str
Partner                  str
Dependents               str
Tenure Months          int64
Phone Service            str
Multiple Lines           str
Internet Service         str
Online Security          str
Online Backup            str
Device Protection        str
Tech Support             str
Streaming TV             str
Streaming Movies         str
Contract                 str
Paperless Billing        str
Payment Method           str
Monthly Charges      float64
Total Charges        float64
Churn Label              str
Churn Value            int64
Churn Score            int64
CLTV                   int64
Churn Reason             str
dtype: object

## Paso 2: Exploración inicial de los datos

Responda las siguientes preguntas:

- ¿Hay valores fuera de rango o imposibles, según el dominio del problema?
- ¿Hay variables que no varían?
- Qué variables *cualitativas* tienen muchos valores únicos (alta cardinalidad)?
- ¿Hay claves primarias o identificadores únicos?

¿Hay valores fuera de rango o imposibles, según el dominio del problema?

vamos a ver todas las columnas del dataframe para ver si cumplen con lo que deberían tener, primero vamos con la columna "CustomerID" se supone que es un identificador único para cada registro, entonces debería tener datos únicos

In [19]:
assert len(df['CustomerID'].unique()) == df.shape[0]

siempre hay datos diferentes en la columna CustomerID, cumple con lo que dice la columna pero no es uitl para el analisis de datos, por lo tanto se elimina

In [20]:
len(df['Count'].unique())

1

En la columna count solo hay un dato. En la practica esto no es útil pero no es raro para el dominio del problema.

In [21]:
len(df["Country"].unique())

1

"Country" es otra columaa con un único dato, pero esto sigue teniendo sentido para nuestro problema, lo mismo para con la columna "State"

In [22]:
len(df["State"].unique())

1

In [26]:
df["City"].unique()

<StringArray>
[    'Los Angeles',   'Beverly Hills', 'Huntington Park',         'Lynwood',
  'Marina Del Rey',       'Inglewood',    'Santa Monica',        'Torrance',
        'Whittier',        'La Habra',
 ...
      'Janesville',      'Litchfield',        'Loyalton',        'Madeline',
    'Markleeville',         'Milford',         'Calpine',        'Standish',
        'Tulelake',  'Olympic Valley']
Length: 1129, dtype: str

La columna "City" contiene ciudades del estado de california de estados unidos, está columna se ve bien

Esta columna describe los códigos postales de la residencias principales de los clientes, tiene sentido que no existan tantos códigos postales difenrentes como registros por que pueden existir clientes que compartan el mismo código postal en la residencia principal

In [40]:
df['Zip Code'].unique()

array([90003, 90005, 90006, ..., 96128, 96134, 96146], shape=(1652,))

esta columan es la combinación de latitud y longitud de la residencia principal de los clientes, tiene sentido que sea una tupla con valores numericos

In [28]:
df['Lat Long'].unique()

<StringArray>
['33.964131, -118.272783',  '34.059281, -118.30742', '34.048013, -118.293953',
 '34.062125, -118.315709', '34.039224, -118.266293', '34.066367, -118.309868',
  '34.02381, -118.156582', '34.066303, -118.435479', '34.099869, -118.326843',
 '34.089953, -118.294824',
 ...
 '40.294034, -120.512622', '40.507272, -120.338228', '39.637471, -120.226338',
 '41.042003, -120.506086',  '38.735789, -119.85798', '40.181279, -120.392967',
 '39.672813, -120.456699', '40.346634, -120.386422', '41.813521, -121.492666',
 '39.191797, -120.212401']
Length: 1652, dtype: str

La latitud puede tomar valores de 0 a 90 grados, como los valores están en ese rango los datos de esta columna parecen tener sentido

In [36]:
df["Latitude"].max(), df["Latitude"].min()

(np.float64(41.962127), np.float64(32.555828))

La longitud puede tener un rango de -180 a 0 grados, como los valores mínimos están en este rango, esto tiene sentido

In [39]:
df["Longitude"].max(), df["Longitude"].min()

(np.float64(-114.192901), np.float64(-124.301372))

En la columna Gender solo hay dos generos, los datos están bien :)

In [41]:
df["Gender"].unique()

<StringArray>
['Male', 'Female']
Length: 2, dtype: str

"Senior Citizen" indica si el cliente es mayor de 65 años o no, y en efecto solo tenemos estos dos valores

In [44]:
df["Senior Citizen"].unique()

<StringArray>
['No', 'Yes']
Length: 2, dtype: str

Esta columna indica si el cliente tiene pareja o no

In [47]:
df["Partner"].unique()

<StringArray>
['No', 'Yes']
Length: 2, dtype: str

Esta columna indica si el cliente vive con alguien dependiente o no

In [49]:
df["Dependents"].unique()

<StringArray>
['No', 'Yes']
Length: 2, dtype: str

¿Hay variables que no varían?

Para solucionar la pregunta se puede hacer de una forma directa

In [56]:
df.nunique() == 1

CustomerID           False
Count                 True
Country               True
State                 True
City                 False
Zip Code             False
Lat Long             False
Latitude             False
Longitude            False
Gender               False
Senior Citizen       False
Partner              False
Dependents           False
Tenure Months        False
Phone Service        False
Multiple Lines       False
Internet Service     False
Online Security      False
Online Backup        False
Device Protection    False
Tech Support         False
Streaming TV         False
Streaming Movies     False
Contract             False
Paperless Billing    False
Payment Method       False
Monthly Charges      False
Total Charges        False
Churn Label          False
Churn Value          False
Churn Score          False
CLTV                 False
Churn Reason         False
dtype: bool

hay tres columnas que solo tienen un dato único, Count, Country y State.

- Qué variables cualitativas tienen muchos valores únicos (alta cardinalidad)?

primero buscamos cuales con mis variables cualitativas, por ahora suponemos que las variables cualitativas son aquellas que su tipo de datos es una cadena de texto, pero como sabemos que la columna "Lat Long" es la combinación de las columnas "Latitude" y "Longitude" la vamos a eliminar para efectos practicos en un nuevo dataframe y asi mismo vamos a hacer con las siguientes columnas:

- CustomerID

In [85]:
new_df = df.drop(columns=["Lat Long"])
new_df = new_df.drop(columns=["CustomerID"])

In [90]:
select_cols = new_df.columns[
    new_df.dtypes.isin(["object", "string"]) & (new_df.nunique() > 2)
]

In [91]:
select_cols

Index(['City', 'Multiple Lines', 'Internet Service', 'Online Security',
       'Online Backup', 'Device Protection', 'Tech Support', 'Streaming TV',
       'Streaming Movies', 'Contract', 'Payment Method', 'Churn Reason'],
      dtype='str')

In [92]:
df_filtered = df[select_cols]

In [93]:
df_filtered.nunique()

City                 1129
Multiple Lines          3
Internet Service        3
Online Security         3
Online Backup           3
Device Protection       3
Tech Support            3
Streaming TV            3
Streaming Movies        3
Contract                3
Payment Method          4
Churn Reason           20
dtype: int64

Lo que acabamos de hacer es utilizar operaciones lógicas para obtener las columans del tipo de datos string que tienen más de dos datos únicos, y nos da como resultado que City, Churn Reason son las columnas con mayor cardinalidad (siendo City la de mayor cardinalidad)

¿Hay claves primarias o identificadores únicos?

In [95]:
df.nunique() == df.shape[0]

CustomerID            True
Count                False
Country              False
State                False
City                 False
Zip Code             False
Lat Long             False
Latitude             False
Longitude            False
Gender               False
Senior Citizen       False
Partner              False
Dependents           False
Tenure Months        False
Phone Service        False
Multiple Lines       False
Internet Service     False
Online Security      False
Online Backup        False
Device Protection    False
Tech Support         False
Streaming TV         False
Streaming Movies     False
Contract             False
Paperless Billing    False
Payment Method       False
Monthly Charges      False
Total Charges        False
Churn Label          False
Churn Value          False
Churn Score          False
CLTV                 False
Churn Reason         False
dtype: bool

Si, customerid es el identificador único.

En el proceso se igualó a `df.shape[0]` por que aquí está la cantidad de registros del dataframe

## Paso 3: Identificación de datos duplicados, datos nulos y problemas de formateo

Descarte las variables que solo toman un valor y las que son clave primaria. También descarte las variables **City** y **Lat Long**.

Responda las siguientes preguntas:

- ¿Qué variables tienen valores nulos? ¿En qué proporción?
- ¿Hay registros duplicados?
- ¿Hay variables con problemas de formateo (fechas, monedas, mayúsculas/minúsculas)?

¿Qué variables tienen valores nulos? ¿En qué proporción?

In [19]:
df.isnull().sum()

City                    0
Zip Code                0
Lat Long                0
Latitude                0
Longitude               0
Gender                  0
Senior Citizen          0
Partner                 0
Dependents              0
Tenure Months           0
Phone Service           0
Multiple Lines          0
Internet Service        0
Online Security         0
Online Backup           0
Device Protection       0
Tech Support            0
Streaming TV            0
Streaming Movies        0
Contract                0
Paperless Billing       0
Payment Method          0
Monthly Charges         0
Total Charges          11
Churn Label             0
Churn Value             0
Churn Score             0
CLTV                    0
Churn Reason         5174
dtype: int64

hay 11 datos nulos en la columna "Total Charges", y 5174 en Churn Reason.
Las proporciones para cada uno son:

In [102]:
f"la proporción de datos nulos para 'Churn Reason' es {(df["Churn Reason"].isnull().sum() / df.shape[0]) * 100}%"

"la proporción de datos nulos para 'Churn Reason' es 73.4630129206304%"

In [103]:
f"la proporción de datos nulos para 'Churn Reason' es {(df["Total Charges"].isnull().sum() / df.shape[0]) * 100}%"

"la proporción de datos nulos para 'Churn Reason' es 0.1561834445548772%"

- ¿Hay registros duplicados?

Sabemos que la columna "CustomerID" tiene datos únicos, esto hace imposible que existan datos duplicados, entonces para intentar datos duplicados vamos a eliminar esta columna en un dataframe diferente

In [104]:
new_df = df.drop("CustomerID", axis=1)

In [105]:
new_df.duplicated().sum()

np.int64(0)

no hay datos duplicados

¿Hay variables con problemas de formateo (fechas, monedas, mayúsculas/minúsculas)?

Si, con anterioridad habiamos visto que la columan "Long Lat" los datos están separados por comas, en este caso no es un problema por que existen columnas independientes que especifican cada parte

## Paso 4: Análisis de variables cuantitativas

Haga análisis de normalidad de las variables cuantitativas. Indique cuáles variables cumplen con el supuesto de normalidad y cuáles no.

Haga un análisis gráfico de outliers, y como resultado de este indique que variables tienen datos atípicos.

## Paso 5: Análisis de variables cualitativas

Responda a las siguientes preguntas:

- ¿Cómo se comportan las variables categóricas?
- ¿Existen variables de alta cardinalidad?
- ¿Existen categorías atípicas?

## Paso 6: Análisis bivariado.

Se va a suponer que con este archivo de datos se desea hacer un modelo de aprendizaje de máquina que prediga la **Churn Value**.


Mediante análisis de gráficas responda las siguientes preguntas:

- ¿Cómo se relaciona cada una de las variables cuantitativas con la variable que se desea predecir?
- ¿Cómo se relaciona cada una de las variables cualitativas con la variable que se desea predecir?
- De acuerdo con estos análisis, ¿cuáles variables parecerían ser buenas predictoras, y cuáles no?
